src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -5864,50 +5986,1026 @@ void Main(string? s)
+{
+    public void Main(string? key)
+    {
+        key = ""hello"";
+        CopyOrDefault(key, out var s);
+        s/*T:string?*/.ToString(); // warn
+    }
+    public static void CopyOrDefault<T>(T key, out T? value) => throw null;
+}
+", parseOptions: TestOptions.Regular8);
+
+            VerifyOutVar(c, "string?");
+            c.VerifyTypes();
+            c.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         s/*T:string?*/.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(8, 9)
+                );
+        }
+
+        [Fact]
+        public void MethodWithGenericNullableArrayOutParameter()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+public class C
+{
+    public void Main(string key)
+    {
+        CopyOrDefault(key, out var s);
+        s/*T:string?[]*/[0].ToString(); // warn
+    }
+    public static void CopyOrDefault<T>(T key, out T?[] value) => throw null;
+}
+", parseOptions: TestOptions.Regular8);
+
+            VerifyOutVar(c, "string?[]");
+            c.VerifyTypes();
+            c.VerifyDiagnostics(
+                // (7,9): warning CS8602: Possible dereference of a null reference.
+                //         s/*T:string?[]*/[0].ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s/*T:string?[]*/[0]").WithLocation(7, 9)
+                );
+        }
+
+        [Fact]
+        public void MethodWithGenericNullableArrayOutParameter_WithNullableArgument()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+public class C
+{
+    public void Main(string? key)
+    {
+        CopyOrDefault(key, out var s);
+        s/*T:string?[]*/[0].ToString(); // warn
+    }
+    public static void CopyOrDefault<T>(T key, out T?[] value) => throw null;
+}
+", parseOptions: TestOptions.Regular8);
+
+            VerifyOutVar(c, "string?[]");
+            c.VerifyTypes();
+            c.VerifyDiagnostics(
+                // (7,9): warning CS8602: Possible dereference of a null reference.
+                //         s/*T:string?[]*/[0].ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s/*T:string?[]*/[0]").WithLocation(7, 9)
+                );
+        }
+
+        [Fact]
+        public void MethodWithGenericArrayOutParameter_WithNonNullableArgument()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+public class C
+{
+    public void Main(string key)
+    {
+        CopyOrDefault(key, out var s);
+        s/*T:string![]*/[0].ToString(); // ok
+    }
+    public static void CopyOrDefault<T>(T key, out T[] value) => throw null;
+}
+", parseOptions: TestOptions.Regular8);
+
+            VerifyOutVar(c, "string![]");
+            c.VerifyTypes();
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void MethodWithGenericArrayOutParameter_WithNonNullArgument()
```

Comment:
`_WithNullableArgument`?

Reply:
This is named correctly. `key` is declared as nullable, but its state is `non-null` when it is used as argument to `CopyOrDefault`.
Contrast that with the two tests above.

---
In reply to: [188460551](https://github.com/dotnet/roslyn/pull/26828#discussion_r188460551) [](ancestors = 188460551)